File: packages/fob/release_ci_exclusions_test.go
package main_test

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	"gopkg.in/yaml.v3"
)

// releasePleaseConfig is the subset of release-please-config.json we need
// to derive the files a release PR is expected to touch.
type releasePleaseConfig struct {
	Packages map[string]releasePleasePackage `json:"packages"`
}

type releasePleasePackage struct {
	ReleaseType string   `json:"release-type"`
	ExtraFiles  []string `json:"extra-files"`
}

// expectedReleaseOutputs derives the complete set of paths release-please
// writes for this repository from release-please-config.json. The set is
// the source of truth for pull_request path filters: every PR-triggered
// workflow must exclude every path here, or release PRs start creating
// action_required runs again.
func expectedReleaseOutputs(cfg releasePleaseConfig) ([]string, error) {
	if len(cfg.Packages) == 0 {
		return nil, fmt.Errorf("release-please-config.json has no packages")
	}

	seen := map[string]struct{}{
		// Manifest is always written for multi and single-package configs.
		".release-please-manifest.json": {},
	}

	for pkgPath, pkg := range cfg.Packages {
		// CHANGELOG.md lives at the package root (repo root for ".").
		changelog := "CHANGELOG.md"
		if pkgPath != "." && pkgPath != "" {
			changelog = filepath.ToSlash(filepath.Join(pkgPath, "CHANGELOG.md"))
		}

		switch pkg.ReleaseType {
		case "go":
			seen[changelog] = struct{}{}
		case "simple":
			seen[changelog] = struct{}{}
			versionTxt := "version.txt"
			if pkgPath != "." && pkgPath != "" {
				versionTxt = filepath.ToSlash(filepath.Join(pkgPath, "version.txt"))
			}
			seen[versionTxt] = struct{}{}
		case "node":
			seen[changelog] = struct{}{}
			packageJSON := "package.json"
			if pkgPath != "." && pkgPath != "" {
				packageJSON = filepath.ToSlash(filepath.Join(pkgPath, "package.json"))
			}
			seen[packageJSON] = struct{}{}
			// Only include package-lock.json when the repo actually has one.
			// pnpm repositories keep a root pnpm-lock.yaml that release-please
			// does not write.
			lock := "package-lock.json"
			if pkgPath != "." && pkgPath != "" {
				lock = filepath.ToSlash(filepath.Join(pkgPath, "package-lock.json"))
			}
			if _, err := os.Stat(lock); err == nil {
				seen[lock] = struct{}{}
			}
		case "":
			return nil, fmt.Errorf("package %q missing release-type", pkgPath)
		default:
			// Unknown types still produce a CHANGELOG at the package root;
			// fail closed by requiring at least that plus extra-files.
			seen[changelog] = struct{}{}
		}

		for _, extra := range pkg.ExtraFiles {
			if extra == "" {
				continue
			}
			seen[filepath.ToSlash(extra)] = struct{}{}
		}
	}

	out := make([]string, 0, len(seen))
	for p := range seen {
		out = append(out, p)
	}
	sort.Strings(out)
	return out, nil
}

func parseWorkflowOn(data []byte) (map[string]any, error) {
	// Unmarshal into map[any]any so a bare `on:` key (YAML 1.1 boolean
	// true) is preserved as bool true rather than being stringified away.
	var raw map[any]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	for _, key := range []any{"on", true} {
		v, ok := raw[key]
		if !ok {
			continue
		}
		return asStringKeyMap(v)
	}
	return nil, fmt.Errorf("workflow has no on: trigger map")
}

func asStringKeyMap(v any) (map[string]any, error) {
	switch typed := v.(type) {
	case map[string]any:
		return typed, nil
	case map[any]any:
		out := make(map[string]any, len(typed))
		for k, val := range typed {
			out[fmt.Sprint(k)] = val
		}
		return out, nil
	default:
		return nil, fmt.Errorf("unexpected on: value type %T", v)
	}
}

// pathFilter describes how a pull_request trigger filters by path.
type pathFilter struct {
	// kind is "none", "ignore", or "allow".
	kind string
	// patterns are the paths-ignore entries (kind=ignore) or paths entries
	// (kind=allow). Negative allow patterns start with '!'.
	patterns []string
}

func pullRequestPathFilter(on map[string]any) (pathFilter, bool, error) {
	prRaw, ok := on["pull_request"]
	if !ok {
		return pathFilter{}, false, nil
	}

	// Bare `pull_request:` (null / nil) means "all PRs, no path filter".
	if prRaw == nil {
		return pathFilter{kind: "none"}, true, nil
	}

	pr, err := asStringKeyMap(prRaw)
	if err != nil {
		// `pull_request: [opened, synchronize]` list form - no path filter.
		return pathFilter{kind: "none"}, true, nil
	}

	ignore, hasIgnore, err := stringListField(pr, "paths-ignore")
	if err != nil {
		return pathFilter{}, true, err
	}
	allow, hasAllow, err := stringListField(pr, "paths")
	if err != nil {
		return pathFilter{}, true, err
	}
	if hasIgnore && hasAllow {
		return pathFilter{}, true, fmt.Errorf("pull_request sets both paths and paths-ignore")
	}
	if hasIgnore {
		return pathFilter{kind: "ignore", patterns: ignore}, true, nil
	}
	if hasAllow {
		return pathFilter{kind: "allow", patterns: allow}, true, nil
	}
	return pathFilter{kind: "none"}, true, nil
}

func stringListField(m map[string]any, key string) ([]string, bool, error) {
	v, ok := m[key]
	if !ok || v == nil {
		return nil, false, nil
	}
	switch typed := v.(type) {
	case []any:
		out := make([]string, 0, len(typed))
		for i, item := range typed {
			s, ok := item.(string)
			if !ok {
				return nil, true, fmt.Errorf("%s[%d] is %T, want string", key, i, item)
			}
			out = append(out, s)
		}
		return out, true, nil
	case []string:
		return append([]string(nil), typed...), true, nil
	default:
		return nil, true, fmt.Errorf("%s has type %T, want list", key, v)
	}
}

// pathExcluded reports whether path is excluded by the filter, matching
// GitHub's documented paths / paths-ignore semantics closely enough for
// exact file paths (no glob expansion beyond "**" / "*" prefix/suffix forms
// used in this fleet). Exact string matches always work; for allow-lists,
// a later negative pattern ("!path") re-excludes after a positive match.
func pathExcluded(filter pathFilter, path string) bool {
	switch filter.kind {
	case "none":
		return false
	case "ignore":
		for _, pat := range filter.patterns {
			if matchGitHubPath(pat, path) {
				return true
			}
		}
		return false
	case "allow":
		// GitHub: workflow runs only if at least one path matches a positive
		// pattern and is not excluded by a later negative pattern.
		// We only need "would this single release-output path cause a run?".
		// It causes a run only if some positive pattern matches it and no
		// subsequent negative pattern excludes it.
		matched := false
		for _, pat := range filter.patterns {
			if strings.HasPrefix(pat, "!") {
				if matched && matchGitHubPath(strings.TrimPrefix(pat, "!"), path) {
					matched = false
				}
				continue
			}
			if matchGitHubPath(pat, path) {
				matched = true
			}
		}
		// Excluded when it does not survive as a positive match.
		return !matched
	default:
		return false
	}
}

// matchGitHubPath is a minimal matcher for the patterns this repository
// uses: exact paths, trailing /**, and simple * / ** globs.
func matchGitHubPath(pattern, path string) bool {
	pattern = filepath.ToSlash(pattern)
	path = filepath.ToSlash(path)
	if pattern == path {
		return true
	}
	// directory/**
	if strings.HasSuffix(pattern, "/**") {
		prefix := strings.TrimSuffix(pattern, "/**")
		if path == prefix || strings.HasPrefix(path, prefix+"/") {
			return true
		}
	}
	// *.ext at any depth is not used here; support **/*.ext for root+nested.
	if strings.HasPrefix(pattern, "**/") {
		suf := strings.TrimPrefix(pattern, "**/")
		if path == suf || strings.HasSuffix(path, "/"+suf) {
			return true
		}
		// **/*.md style
		if strings.HasPrefix(suf, "*.") {
			ext := strings.TrimPrefix(suf, "*")
			if strings.HasSuffix(path, ext) {
				return true
			}
		}
	}
	if strings.HasPrefix(pattern, "*.") {
		return strings.HasSuffix(path, strings.TrimPrefix(pattern, "*"))
	}
	return false
}

func TestPullRequestWorkflowsExcludeReleasePleaseOutputs(t *testing.T) {
	cfgBytes, err := os.ReadFile("release-please-config.json")
	if err != nil {
		t.Fatalf("read release-please-config.json: %v", err)
	}
	var cfg releasePleaseConfig
	if err := json.Unmarshal(cfgBytes, &cfg); err != nil {
		t.Fatalf("parse release-please-config.json: %v", err)
	}

	expected, err := expectedReleaseOutputs(cfg)
	if err != nil {
		t.Fatalf("derive release outputs: %v", err)
	}
	if len(expected) == 0 {
		t.Fatal("derived empty release output set")
	}
	t.Logf("expected release-please outputs: %v", expected)

	entries, err := os.ReadDir(filepath.Join(".github", "workflows"))
	if err != nil {
		t.Fatalf("read .github/workflows: %v", err)
	}

	var prWorkflows int
	for _, ent := range entries {
		if ent.IsDir() {
			continue
		}
		name := ent.Name()
		if !strings.HasSuffix(name, ".yml") && !strings.HasSuffix(name, ".yaml") {
			continue
		}
		path := filepath.Join(".github", "workflows", name)
		data, err := os.ReadFile(path)
		if err != nil {
			t.Fatalf("read %s: %v", path, err)
		}
		on, err := parseWorkflowOn(data)
		if err != nil {
			t.Fatalf("parse on: in %s: %v", path, err)
		}
		filter, hasPR, err := pullRequestPathFilter(on)
		if err != nil {
			t.Fatalf("%s: %v", path, err)
		}
		if !hasPR {
			continue
		}
		prWorkflows++

		var missing []string
		for _, rel := range expected {
			if !pathExcluded(filter, rel) {
				missing = append(missing, rel)
			}
		}
		if len(missing) > 0 {
			t.Errorf("%s pull_request filter must exclude every release-please output; missing: %s",
				path, strings.Join(missing, ", "))
		}
	}

	if prWorkflows == 0 {
		t.Fatal("no pull_request-triggered workflows found under .github/workflows")
	}
}

func TestExpectedReleaseOutputsIncludesConfiguredExtraFiles(t *testing.T) {
	// Sanity: the derivation must surface flake.nix from this repo's config
	// so a future extra-files addition cannot silently drop out of the guard.
	cfgBytes, err := os.ReadFile("release-please-config.json")
	if err != nil {
		t.Fatal(err)
	}
	var cfg releasePleaseConfig
	if err := json.Unmarshal(cfgBytes, &cfg); err != nil {
		t.Fatal(err)
	}
	got, err := expectedReleaseOutputs(cfg)
	if err != nil {
		t.Fatal(err)
	}
	want := []string{".release-please-manifest.json", "CHANGELOG.md", "flake.nix"}
	if strings.Join(got, ",") != strings.Join(want, ",") {
		t.Fatalf("expected %v, got %v", want, got)
	}
}

File: packages/tasks-axi/test/release-ci-exclusions.test.ts
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { parse } from "yaml";

const root = fileURLToPath(new URL("..", import.meta.url));
const workflowsDir = join(root, ".github", "workflows");

/**
 * Derive the exact release-please output set from config + workflow inputs.
 * Keep this aligned with the fleet audit rule in firstmate's release-please CI
 * report: node -> package.json (+ package-lock.json if present), changelog,
 * extra-files, and the manifest path.
 */
function expectedReleaseOutputs(): string[] {
  const config = JSON.parse(
    readFileSync(join(root, "release-please-config.json"), "utf8"),
  ) as {
    "release-type"?: string;
    "changelog-path"?: string;
    "version-file"?: string;
    "extra-files"?: Array<string | { path?: string }>;
    packages?: Record<
      string,
      {
        "release-type"?: string;
        "changelog-path"?: string;
        "version-file"?: string;
        "extra-files"?: Array<string | { path?: string }>;
      }
    >;
  };
  const pkg = config.packages?.["."] ?? {};
  const releaseType = pkg["release-type"] ?? config["release-type"] ?? "node";
  const changelog =
    pkg["changelog-path"] ?? config["changelog-path"] ?? "CHANGELOG.md";

  const expected: string[] = [changelog];
  switch (releaseType) {
    case "simple":
      expected.push(
        pkg["version-file"] ?? config["version-file"] ?? "version.txt",
      );
      break;
    case "node":
      expected.push("package.json");
      if (existsSync(join(root, "package-lock.json"))) {
        expected.push("package-lock.json");
      }
      break;
    case "go":
      break;
    default:
      throw new Error(
        `unsupported release-please release-type for ignore derivation: ${releaseType}`,
      );
  }

  const extra = pkg["extra-files"] ?? config["extra-files"] ?? [];
  for (const entry of extra) {
    const path = typeof entry === "string" ? entry : entry?.path;
    if (path) expected.push(path);
  }

  let manifest = ".release-please-manifest.json";
  const releaseWorkflow = readFileSync(
    join(workflowsDir, "release-please.yml"),
    "utf8",
  );
  const manifestMatch = releaseWorkflow.match(/manifest-file:\s*(\S+)/);
  if (manifestMatch) manifest = manifestMatch[1];
  expected.push(manifest);

  return [...new Set(expected)];
}

function loadWorkflowOn(filePath: string): Record<string, unknown> | null {
  const doc = parse(readFileSync(filePath, "utf8")) as
    | Record<string | boolean, unknown>
    | null;
  if (!doc || typeof doc !== "object") return null;
  // YAML 1.1 may parse a bare `on:` key as boolean true.
  const on = doc.on ?? doc[true];
  if (!on || typeof on !== "object" || Array.isArray(on)) return null;
  return on as Record<string, unknown>;
}

type PullRequestFilter =
  | { kind: "unfiltered" }
  | { kind: "paths-ignore"; paths: string[] }
  | { kind: "paths"; paths: string[] };

function pullRequestFilterCoverage(pr: unknown): PullRequestFilter {
  if (pr == null) {
    return { kind: "unfiltered" };
  }
  if (typeof pr !== "object" || Array.isArray(pr)) {
    // `pull_request:` bare form means no path filter.
    return { kind: "unfiltered" };
  }

  const record = pr as Record<string, unknown>;
  if (Array.isArray(record["paths-ignore"])) {
    return {
      kind: "paths-ignore",
      paths: record["paths-ignore"].map(String),
    };
  }

  if (Array.isArray(record.paths)) {
    return { kind: "paths", paths: record.paths.map(String) };
  }

  return { kind: "unfiltered" };
}

function globMatch(pattern: string, path: string): boolean {
  // Minimal support for the `**` / `*` patterns used in workflow path filters.
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*\*/g, "::DOUBLE::")
    .replace(/\*/g, "[^/]*")
    .replace(/::DOUBLE::/g, ".*");
  return new RegExp(`^${escaped}$`).test(path);
}

function isCovered(filter: PullRequestFilter, releasePath: string): boolean {
  if (filter.kind === "unfiltered") return false;

  if (filter.kind === "paths-ignore") {
    return filter.paths.includes(releasePath);
  }

  // paths allow-list: a release path is "covered" (will not create a run on its
  // own) when no positive pattern matches it, or a later negation excludes it.
  let matched = false;
  for (const pattern of filter.paths) {
    if (pattern.startsWith("!")) {
      const negated = pattern.slice(1);
      if (
        matched &&
        (negated === releasePath || globMatch(negated, releasePath))
      ) {
        matched = false;
      }
      continue;
    }
    if (pattern === releasePath || globMatch(pattern, releasePath)) {
      matched = true;
    }
  }
  // Covered means the path does NOT cause the workflow to run.
  return !matched;
}

describe("release-please CI exclusions", () => {
  const expected = expectedReleaseOutputs();

  it("derives the node release-output set for this repository", () => {
    expect(expected).toEqual([
      "CHANGELOG.md",
      "package.json",
      ".release-please-manifest.json",
    ]);
  });

  it("every pull_request workflow ignores the full release-output set", () => {
    const files = readdirSync(workflowsDir).filter((name) =>
      name.endsWith(".yml"),
    );
    const prWorkflows: { name: string; filter: PullRequestFilter }[] = [];

    for (const name of files) {
      const filePath = join(workflowsDir, name);
      const on = loadWorkflowOn(filePath);
      if (!on || !("pull_request" in on)) continue;
      prWorkflows.push({
        name,
        filter: pullRequestFilterCoverage(on.pull_request),
      });
    }

    expect(prWorkflows.map((w) => w.name).sort()).toEqual([
      "ci.yml",
      "guard-generated-files.yml",
      "no-mistakes-required.yml",
    ]);

    const failures: string[] = [];
    for (const { name, filter } of prWorkflows) {
      const missing = expected.filter((path) => !isCovered(filter, path));
      if (missing.length > 0) {
        failures.push(`${name} missing coverage for: ${missing.join(", ")}`);
      }
    }

    expect(failures).toEqual([]);
  });

  it("does not attach path filters to non-pull_request triggers on ci.yml", () => {
    const on = loadWorkflowOn(join(workflowsDir, "ci.yml"));
    expect(on).not.toBeNull();
    expect(on!.push).toEqual({ branches: ["main"] });
    const pr = on!.pull_request as Record<string, unknown>;
    expect(pr.branches).toEqual(["main"]);
    expect(pr["paths-ignore"]).toEqual([
      ".release-please-manifest.json",
      "CHANGELOG.md",
      "package.json",
    ]);
    expect(on!.release).toBeUndefined();
    expect(on!.workflow_dispatch).toBeUndefined();
  });

  it("keeps bot author exemptions on guard and no-mistakes jobs", () => {
    const guard = readFileSync(
      join(workflowsDir, "guard-generated-files.yml"),
      "utf8",
    );
    const nmr = readFileSync(
      join(workflowsDir, "no-mistakes-required.yml"),
      "utf8",
    );
    expect(guard).toContain("github-actions[bot]");
    expect(guard).toContain("release-please[bot]");
    expect(nmr).toContain("github-actions[bot]");
    expect(nmr).toContain("dependabot[bot]");
    expect(nmr).toContain("release-please[bot]");
  });
});
